util.cc tests

Moliholy opened the pull request:
It includes the fix in fs_traversal.h so that I can remove sockets from the directory where I am working.

Moliholy commented on cvmfs/util.cc:
```diff
@@ -1125,6 +1095,9 @@ string ReplaceAll(const string &haystack, const string &needle,
   size_t pos = 0;
   const unsigned needle_size = needle.size();
+  if(needle == "")
+    return result;
```
Otherwise it enters in an infinite loop

jblomer commented:
Looks good in a first skim!

Now, before we go into details, there are the usual things before submitting a pull request:
1. This one you'll most likely figure out yourself
2. At least on my machine, the T_Util unit tests do not run but produce the error

    [ RUN      ] T_Util.BlockSignal
    User defined signal 1

Regarding the `fs_traversal.h`, I'll close the other (pull request)[https://github.com/cvmfs/cvmfs/pull/884] but please take the comments into account and make the list of supported file types complete including `S_ISBLK` and `S_ISFIFO`

reneme commented on cvmfs/util.cc:
```diff
@@ -927,15 +906,6 @@ bool HasSuffix(
 }
 
 
-bool IsNumeric(const std::string &str) {
```
In a currently pending [Pull Request](https://github.com/cvmfs/cvmfs/pull/885) I am actually [using this function](https://github.com/cvmfs/cvmfs/pull/885/files#diff-e97df2d1b1380e814faaf512cc866e46R142). Please don't remove it.

jblomer replied:
I think we have this already in the sanitizer classes. Wouldn't it make more sense to take it from there?

reneme replied:
Ah, sweet... I didn't know about that thing. Jep, I'm going to switch to that then.

jblomer replied:
Should be removed from the header, too

reneme commented on test/unittests/t_util.cc:
```diff
@@ -209,3 +307,853 @@ TEST(T_Util, IsHttpUrl) {
+  close(fd_stderr);
+}
+
+TEST_F(T_Util, ManagedExec) {
```
To my own little surprise, I found [this](https://github.com/cvmfs/cvmfs/blob/devel/test/unittests/t_managed_exec.cc). O.o I actually jotted together a couple of unit tests for `ManagedExec` back in 2013.

jblomer replied:
Very well, @moliholy can you make these unit tests part of t_util.cc?

Moliholy replied:
Ops, I skipped this comment, I'll do it right now!

jblomer commented:
I think we have this already in the sanitizer classes.  Wouldn't it make more sense to take it from there?

Moliholy commented:
Updated to the latest version of devel, including René's PR (https://github.com/cvmfs/cvmfs/pull/906).
Now all tests pass.

Moliholy commented:
Passed cpplint through fs_traversal.h

jblomer commented on cvmfs/util.cc:
```diff
@@ -822,13 +823,6 @@ string StringifyTimeval(const timeval value) {
 }
 
 
-string StringifyIpv4(const uint32_t ip4_address) {
```
Should be removed from the header, too

jblomer commented on cvmfs/util.cc:
```diff
@@ -875,21 +869,6 @@ uint64_t String2Uint64(const string &value) {
 }
 
 
-uint64_t HexString2Uint64(const string &value) {
```
Should be removed from the header, too

jblomer commented on cvmfs/util.cc:
```diff
@@ -875,21 +869,6 @@ uint64_t String2Uint64(const string &value) {
-}
-
-
-int HexDigit2Int(const char digit) {
```
Should be removed from the header, too

jblomer commented on test/unittests/t_util.cc:
```diff
@@ -32,7 +38,94 @@ class ThreadDummy {
+class T_Util : public ::testing::Test {
+ protected:
+  virtual void SetUp() {
+    const bool retval = MkdirDeep(string(sandbox), 0700);
```
It would be better to use CreateTempDir here for the sandbox, which ensures a fresh directory is used

Moliholy replied:
Funny thing I've realized right now when using CreateTempDir: it doesn't use the second parameter. Temporary directories are always created using by default 0700 permissions.

jblomer commented on test/unittests/t_util.cc:
```diff
@@ -32,7 +38,94 @@ class ThreadDummy {
+
+ protected:
+  static void WriteBuffer(int fd) {
+    sleep(1);
```
Is this required?  Can it be replaced by `SafeSleepMs`?

Moliholy replied:
Yes, it could. I was trying not to use functions that are going to be tested in order to avoid bug creations, but with this one there shouldn't be any problem.

jblomer commented on test/unittests/t_util.cc:
```diff
@@ -32,7 +38,94 @@ class ThreadDummy {
+    kill(getpid(), signal);
+  }
+
+  static const char sandbox[];
```
This one and the following don't need to be static, do they?  Can't this just be strings that are initialized in the test fixture setup?

Moliholy replied:
Yes, of course they can. However they will always be initialized the same way and remain constant... so just for performance I thought to put them as constant strings, but this triggered warnings from cpplint, so I changed them to const char[]. But yes, they could perfectly be const string attributes, which maybe would make things easier in the rest of the code, avoiding many conversions to string instances

jblomer commented on test/unittests/t_util.cc:
```diff
@@ -215,3 +314,855 @@ TEST(T_Util, IsHttpUrl) {
+
+TEST_F(T_Util, WritePipe) {
+  int fd[2];
+  void *buffer_output = calloc(20, sizeof(char));
```
Here and for the other uses of `calloc`, better use scalloc from smalloc.h

Moliholy replied:
Fixed using scalloc.

jblomer commented on test/unittests/t_util.cc:
```diff
@@ -215,3 +314,855 @@ TEST(T_Util, IsHttpUrl) {
+  EXPECT_LE(0, fd = LockFile(filename));
+
+  tbb::tbb_thread thread(LockFileTest, filename, &retval);
+  sleep(2);
```
That's a bit long.  Use `SafeSleepMs` with 100ms.  I don't see a way to avoid the waiting altogether, unfortunately.  We should consider adding a `#define CVMFS_TESTING` to cvmfs.

reneme replied:
Couldn't you just put a `Future<bool>` into LockFileTest that gets filled once the thread has started executing and put a `Future<bool>::Get()` right under the thread creation here. Or am I missing something?

Moliholy replied:
You don't want to release the main thread when the other is active, but when the other is blocked in the function we are testing here. But if the other is blocked... how is he going to tell the main thread that he actually is? They both have to wait until each other is ready, but at the same time there is no way to warn them.

jblomer commented on test/unittests/t_util.cc:
```diff
@@ -215,3 +314,855 @@ TEST(T_Util, IsHttpUrl) {
+  void *buffer_output = calloc(20, sizeof(char));
+  MakePipe(fd);
+
+  tbb::tbb_thread writer(WriteBuffer, fd[1]);
```
What happens if you write without sleeping?

Moliholy replied:
Inside WriteBuffer it waits, but in the other thread.
If it wouldn't wait then the writer could write everything before the reader reads it. The idea of the ReadHalfPipe function is to read something from the pipe even if the writer hasn't started to write yet. For that there _must_ be something in the pipe, and the idea of this test was to check what happens if there is nothing written. Of course it indefinitely waits until the writer (atomically) flushes what it has written, and then the reader consumes it.

jblomer commented on test/unittests/t_util.cc:
```diff
@@ -215,3 +314,855 @@ TEST(T_Util, IsHttpUrl) {
+TEST_F(T_Util, DiffTimeSeconds) {
+  struct timeval start = CreateTimeval(2, 900000);
+  struct timeval end = CreateTimeval(3, 200000);
+  EXPECT_EQ(0.3, DiffTimeSeconds(start, end));
```
There are special macros for comparing floating point numbers.  Due to rounding errors, you can never by sure about the exact result of floating point arithmetic.

Moliholy replied:
Updated using EXPECT_DOUBLE_EQ macro. Thanks for the warning!

jblomer commented on test/unittests/t_util.cc:
```diff
@@ -215,3 +314,855 @@ TEST(T_Util, IsHttpUrl) {
+
+  start = CreateTimeval(8, 400000);
+  end = CreateTimeval(1, 800000);
+  EXPECT_NE(-6.6, DiffTimeSeconds(start, end));
```
What is this supposed to test?

Moliholy replied:
That the difference is not negative. In principle one would expect to find a negative difference, but the functions is not supposed to return a negative value. However it could be changed for EXPECT_GT(0, DiffTimeSeconds(start, end)

jblomer commented on test/unittests/t_util.cc:
```diff
@@ -32,7 +38,94 @@ class ThreadDummy {
+
+  static void SendSignal(int signal) {
+    BlockSignal(signal);
+    sleep(1);
```
It's risky to rely on sleep.  Try to figure out if you find an order of events and threads that doesn't require sleep.

Moliholy replied:
Fixed the way we mentioned before

jblomer commented on test/unittests/t_util.cc:
```diff
@@ -215,3 +314,855 @@ TEST(T_Util, IsHttpUrl) {
+  ASSERT_LT(0.0, sw.GetTime());
+  ASSERT_DEATH(sw.Stop(), ".*");
+  sw.Reset();
+  ASSERT_EQ(0.0, sw.GetTime());
```
See above on floating point comparison

jblomer commented:
Looks quite good!  Please check the minor comments.

Moliholy commented:
Updated.
I have removed most of the "sleeps" because I found other ways to implement it (for signal blocking concretely) or because they were not really necessary (ReadHalfPipe).
Unittests and cpplint passed

Moliholy commented:
Removed t_managed_exec.cc. Its tests have been added to t_util.cc.
Cpplint and unittests successfully passed

jblomer commented on cvmfs/fs_traversal.h:
```diff
@@ -182,6 +185,18 @@ class FileSystemTraversal {
+      } else if (S_ISBLK(info.st_mode)) {
+        LogCvmfs(kLogFsTraversal, kLogVerboseMsg,
+            "passing blocked device %s/%s", path.c_str(), dit->d_name);
+        Notify(fn_new_socket, path, dit->d_name);
```
this should not call `fn_new_socket` but `fn_new_block`.

jblomer commented on cvmfs/fs_traversal.h:
```diff
@@ -182,6 +185,18 @@ class FileSystemTraversal {
+      } else if (S_ISFIFO(info.st_mode)) {
+        LogCvmfs(kLogFsTraversal, kLogVerboseMsg, "passing FIFO %s/%s",
+                 path.c_str(), dit->d_name);
+        Notify(fn_new_socket, path, dit->d_name);
```
Same as above, but for fifo.  Please also add test cases that catch this problem.

Moliholy commented:
fs_traversal modified and the corresponding tests have been added.
Cpplint and unitests passed.

jblomer commented on test/unittests/t_fs_traversal.cc:
```diff
@@ -26,7 +27,10 @@ class T_FsTraversal : public ::testing::Test {
-      Unspecified
+      Unspecified,
+      Socket,
+      BlockedDevice,
```
Not a show stopper for the merging, but it is actually a "block device" (because data is transferred in blocks, it's not blocked).

jblomer commented on test/unittests/t_util.cc:
```diff
@@ -215,3 +328,960 @@ TEST(T_Util, IsHttpUrl) {
+
+TEST_F(T_Util, SafeSleepMs) {
+  StopWatch sw;
+  unsigned time = 1000;
```
This can be decreased to 50 so that the unit tests run faster.  I'll do it during merging.

Moliholy commented:
Renamed to block_dev instead of blocked_dev
